src/timer/mod.rs

```diff
@@ -174,6 +172,15 @@ macro_rules! timers {
 
                     self.tim.psc.write(|w| unsafe { w.psc().bits(psc as u16) });
                     self.tim.arr.write(|w| unsafe { w.bits(arr) });
+
+                    // Generate an update event so that PSC and ARR values are copied into their
+                    // shadow registers.
+                    self.tim.egr.write(|w| w.ug().set_bit());
+                    // Setting UG causes UIF to be set after an unspecified number of cycles.
+                    // Wait for it to be set and then clear it, to avoid a spurious interrupt.
+                    while self.tim.sr.read().uif().bit_is_clear() {}
```

Comment:
Hmm on second thought, doing this is probably not a good idea -- whether UIF eventually gets set will depend on a few other bits too. I think maybe my application is accidentally leaving UIE set before starting the timer, which might be affecting the behaviour I am seeing.

Reply:
Let us know when you find out something.

Reply:
@danc86 do you have any updates on this?

Reply:
Sorry it took me so long to get back to this.

I looked at this again yesterday. Back when I sent this PR I was testing it by doing the register manipulations directly in my application, and I was confusing myself because everything appeared to break in weird ways -- but it was just because I had missed the code to enable the peripheral in RCC. :facepalm: 

I went back and again studied RM0444 and checked the behaviour of various combinations of register settings in my application. I found that (assuming the peripheral is enabled in RCC) setting UG does indeed cause UIF to be set by hardware after a few cycles -- regardless of UIE or CEN bits. However this only happens if URS is *not* set. Relevant section of RM0444 says:

> In addition, if the URS bit (update request selection) in
> TIMx_CR1 register is set, setting the UG bit generates an update event UEV but without
> setting the UIF flag (thus no interrupt or DMA request is sent).

The loop I added here will indeed hang the application if URS is already set before this point (which is possible with multiple calls to `.start()`). It also means there is a simpler solution -- set URS before UG. That way, the update even will happen, PSC and ARR values will take effect immediately, and no interrupt will be generated.